#### Here's a copy of my docker-compose.yml

```
---
version: '2'
services:
  zookeeper:
    image: confluentinc/cp-zookeeper:latest
    environment:
      ZOOKEEPER_CLIENT_PORT: 32181
      ZOOKEEPER_TICK_TIME: 2000
    expose:
      - "2181"
      - "2888"
      - "32181"
      - "3888"

  kafka:
    image: confluentinc/cp-kafka:latest
    depends_on:
      - zookeeper
    environment:
      KAFKA_BROKER_ID: 1
      KAFKA_ZOOKEEPER_CONNECT: zookeeper:32181
      KAFKA_ADVERTISED_LISTENERS: PLAINTEXT://kafka:29092
      KAFKA_OFFSETS_TOPIC_REPLICATION_FACTOR: 1
    expose:
      - "9092"
      - "29092"

  cloudera:
    image: midsw205/cdh-minimal:latest
    expose:
      - "8020" # nn
      - "50070" # nn http

  spark:
    image: midsw205/spark-python:0.0.5
    ports:
    - "8888:8888"
    stdin_open: true
    tty: true
    volumes:
      - ~/w205:/w205
    command: bash
    depends_on:
      - cloudera
    environment:
      HADOOP_NAMENODE: cloudera

  mids:
    image: midsw205/base:latest
    stdin_open: true
    tty: true
    volumes:
      - ~/w205:/w205
```

### Overview of docker-compose.yml
The docker-compose.yml defines how multiple containers run and interact with one another. The innermost indent after services lists the containers/services that will be launched and coordinated. Some of the `ports` and `expose` instructions define how the containers communicate with one another while other define how they communicate with other services.   

- **zookeeper:**Is a service that, in our case, coordinates the use of kafka's topics and brokers. When we create a topic in kafka we register it in zookeeper because kafka doesn't have an overview of topic and broker relationships itself. kafka communicates with zookeeper on port 32181.
   

- **kafka:** kafka is a distributed system that Doris analogized to Slack. kafka can intake messages ("publish" or "produce" to kafka) to "topics" we define. The messages are placed in a queue associated with that topic. Once there, the data can be read ("consumed") by subscribers to those topics. In our case, spark is a subscriber to the topics we're creating. In the instructions that follow you'll see messages written to ('produced') and read from ('consumed') on port 29092, which is the port exposed in the above yml file under kafka for that purpose.  

  
- **cloudera:** In our case, cloudera is providing us with a Hadoop instance. Hadoop is a distributed data store that we'll write our data to.   

  
- **spark:** Spark is the analytics engine we use for this project. It's the tool within which we unpack the nested assessment json and where we run our SQL queries. We've modified our docker-compose file to allow jupyter notebooks to communicate with spark on port 8888. We've also mapped, under 'volumes', the virtual machine's w205 folder to a corresponding one inside the spark container.  


- **mids:** Services like the jupyter notebooks and kafkacat are provided by the mids image. Similar to the spark container, we've mapped the virtual machine's w205 folder to a corresponding one inside the mids container. 



#### Step-by-Step of getting data from the json and into kafka, all of which takes place on the commmand line and is in the history.txt file


#### Spin up the cluster using the information in the docker-compose.yml file.

 ```docker-compose up -d```

#### Check out hadoop

```docker-compose exec cloudera hadoop fs -ls /tmp/```

#### Create an assessments topic in kafka. Let zookeeper know . 

```docker-compose exec kafka kafka-topics --create --topic assessments --partitions 1 --replication-factor 1 --if-not-exists --zookeeper zookeeper:32181```

#### Pipe assessments.json into jq. Use the iterator command [] along with the switch -c to send individual elements of the nested JSON into kafkacat. kafkacat will take those lines and produce/publish those lines into kafka using the port 29092 to the topic assessments we create above.

```docker-compose exec mids bash -c "cat /w205/project-2-FuriousGeorge19/assessments.json | jq '.[]' -c | kafkacat -P -b kafka:29092 -t assessments"```

#### Run the following command to grant a previously created Jupyter notebook access to Spark through port 8888. We modified the docker-compose.yml to allow access through this port. 

```docker-compose exec spark env PYSPARK_DRIVER_PYTHON=jupyter PYSPARK_DRIVER_PYTHON_OPTS='notebook --no-browser --port 8888 --ip 0.0.0.0 --allow-root --notebook-dir=/w205/' pyspark```

#### Switch over to the notebook

#### The prior command has made pyspark available in this notebook. 

#### import libraries to enable analysis

In [ ]:
import json
from pyspark.sql import Row
from pyspark.sql import functions as F 
from pyspark.sql.functions import from_json, col, lit, countDistinct, avg, col
from pyspark.sql.types import StructType, StructField, StringType, BooleanType, LongType
import sys 
from pyspark.sql.window import Window

#### Read assessments data from kafka into a pyspark dataframe

In [ ]:
raw_assessments = spark.read.format("kafka").option("kafka.bootstrap.servers", "kafka:29092").option("subscribe","assessments").option("startingOffsets", "earliest").option("endingOffsets", "latest").load() 

In [ ]:
type(raw_assessments)

#### Cache the dataframe to cut back on warnings and make queries faster

In [ ]:
raw_assessments.cache()

#### Check count to make sure data was written properly

In [179]:
raw_assessments.count()

3280

#### Cast to strings

In [ ]:
assessments = raw_assessments.select(raw_assessments.value.cast('string'))

#### Extract json fields

In [ ]:
extracted_assessments = spark.read.json(assessments.rdd.map(lambda x: x.value))



#### Create a temp table

In [ ]:
extracted_assessments.registerTempTable('assessments_df')

#### Look at the structure by looking at the schema:

In [176]:
assessments_df.printSchema()

root
 |-- base_exam_id: string (nullable = true)
 |-- certification: string (nullable = true)
 |-- exam_name: string (nullable = true)
 |-- keen_created_at: string (nullable = true)
 |-- keen_id: string (nullable = true)
 |-- keen_timestamp: string (nullable = true)
 |-- max_attempts: string (nullable = true)
 |-- sequences: struct (nullable = true)
 |    |-- attempt: long (nullable = true)
 |    |-- counts: struct (nullable = true)
 |    |    |-- all_correct: boolean (nullable = true)
 |    |    |-- correct: long (nullable = true)
 |    |    |-- incomplete: long (nullable = true)
 |    |    |-- incorrect: long (nullable = true)
 |    |    |-- submitted: long (nullable = true)
 |    |    |-- total: long (nullable = true)
 |    |    |-- unanswered: long (nullable = true)
 |    |-- id: string (nullable = true)
 |    |-- questions: array (nullable = true)
 |    |    |-- element: struct (containsNull = true)
 |    |    |    |-- id: string (nullable = true)
 |    |    |    |-- options: arra

#### Only sequences is nested. But it is really, really nested and will take a bit of work to pick apart and understand. 

#### Look at unnested columns first

In [83]:
spark.sql("select base_exam_id, certification, exam_name, keen_created_at from assessments_df limit 5").show()

+--------------------+-------------+--------------------+------------------+
|        base_exam_id|certification|           exam_name|   keen_created_at|
+--------------------+-------------+--------------------+------------------+
|37f0a30a-7464-11e...|        false|Normal Forms and ...| 1516717442.735266|
|37f0a30a-7464-11e...|        false|Normal Forms and ...| 1516717377.639827|
|4beeac16-bb83-4d5...|        false|The Principles of...| 1516738973.653394|
|4beeac16-bb83-4d5...|        false|The Principles of...|1516738921.1137421|
|6442707e-7488-11e...|        false|Introduction to B...| 1516737000.212122|
+--------------------+-------------+--------------------+------------------+



In [84]:
spark.sql("select keen_id, keen_timestamp, max_attempts, started_at, user_exam_id from assessments_df limit 5").show()

+--------------------+------------------+------------+--------------------+--------------------+
|             keen_id|    keen_timestamp|max_attempts|          started_at|        user_exam_id|
+--------------------+------------------+------------+--------------------+--------------------+
|5a6745820eb8ab000...| 1516717442.735266|         1.0|2018-01-23T14:23:...|6d4089e4-bde5-4a2...|
|5a674541ab6b0a000...| 1516717377.639827|         1.0|2018-01-23T14:21:...|2fec1534-b41f-441...|
|5a67999d3ed3e3000...| 1516738973.653394|         1.0|2018-01-23T20:22:...|8edbc8a8-4d26-429...|
|5a6799694fc7c7000...|1516738921.1137421|         1.0|2018-01-23T20:21:...|c0ee680e-8892-4e6...|
|5a6791e824fccd000...| 1516737000.212122|         1.0|2018-01-23T19:48:...|e4525b79-7904-405...|
+--------------------+------------------+------------+--------------------+--------------------+



#### We'll do a little exploration to see what these fields might mean. I'll start w/base_exam_id and exam_name. 

In [ ]:
spark.sql("select base_exam_id, exam_name from assessments_df limit 10").show(10, False)

#### base_exam_id appears like it might be the unique key associated with exam_name 

In [191]:
assessments_df.select(countDistinct("exam_name")).show()

+-------------------------+
|count(DISTINCT exam_name)|
+-------------------------+
|                      103|
+-------------------------+



In [192]:
assessments_df.select(countDistinct("base_exam_id")).show()

+----------------------------+
|count(DISTINCT base_exam_id)|
+----------------------------+
|                         107|
+----------------------------+



#### With more time and a stronger knowledge of sql I'd select for exam_name's that matched each other but exam_id's that did not. But using groupby and scanning was quicker in this case as it's only 107 rows long. 

In [85]:
temp_df = assessments_df.select('base_exam_id','exam_name').groupby('base_exam_id','exam_name').count()

In [86]:
temp_df.sort("exam_name", ascending = False).show(107, False)

+------------------------------------+-----------------------------------------------------------------------+-----+
|base_exam_id                        |exam_name                                                              |count|
+------------------------------------+-----------------------------------------------------------------------+-----+
|f432e2e3-7e3a-4a78-b408-49cab5d1fbeb|Working with Algorithms in Python                                      |14   |
|e9b58c58-bf2e-4bde-be81-52fb02ebc892|What's New in JavaScript                                               |2    |
|98a4b6fd-7460-11e6-b6c2-a4d18ccf3cb4|Web & Native Working Together                                          |8    |
|44d5ca66-7462-11e6-9d9f-a8667f27e5dc|View Updating                                                          |4    |
|30fd83f9-4937-4a00-bcf3-42e495fecd55|Using Web Components                                                   |3    |
|b71d9e1e-7525-11e6-927c-a4d18ccf3cb4|Using Storytelling to Effe

#### It turns out 4 exams show up twice in exam_name but with unique ID's. Those are 
- Introduction to Python  
- Great Bash  
- Architectural Considerations for Hadoop Applications  
- Being a Better Introvert

I'd include both columns in the dataset, but provide some notes in a README alerting analysts to the fact that 'exam_name' and 'base_exam_id' are almost, but not quite, 1 to 1 matches and ask whether, if they are indeed meant to be unique, if they should be consolidated under a single id. 

In [ ]:
assessments_df.select(countDistinct("user_exam_id")).show()

#### We'll perform a similar exercise with user_exam_id.
#### We'll start by looking to see if the user_exam_id's are unique, knowing there are 3280 rows in the dataframe.

In [180]:
assessments_df.select(countDistinct("user_exam_id")).show()

+----------------------------+
|count(DISTINCT user_exam_id)|
+----------------------------+
|                        3242|
+----------------------------+



#### Not quite unique. But pretty close. I'll take a look to see if the discrepancy is due to NULLs

In [181]:
assessments_df.filter(assessments_df["user_exam_id"].isNull()).select('user_exam_id').count()

0

#### The discrepancy doesn't look like it's due to NULLs, but potentially due to repeats. As this is likely 'the most unique' of the IDs, if I were to split this table, user_exam_id would be the most likely candidate to serve as a common key between the tables. Given more time, I'd want to dig into why there are repeats, if they should be renumbered and retained or if they should be dropped from the dataset. 

#### The last potential candidate as a unique key is keen_id. We'll take a quick look. 

In [193]:
assessments_df.select(countDistinct("keen_id")).show()

+-----------------------+
|count(DISTINCT keen_id)|
+-----------------------+
|                   3242|
+-----------------------+



#### This has the same number of unique id's as user_exam_id. As I have a clearer understanding of what user_exam_id is than keen_id, I'll use the former as a key if I were to split the table at some point. 

#### I'll move on to looking at the data nested under 'sequences'

I'll print the part of the schema associated with sequences so we don't have to scroll so far up to review it.

In [79]:
assessments_df.select('sequences').printSchema()

root
 |-- sequences: struct (nullable = true)
 |    |-- attempt: long (nullable = true)
 |    |-- counts: struct (nullable = true)
 |    |    |-- all_correct: boolean (nullable = true)
 |    |    |-- correct: long (nullable = true)
 |    |    |-- incomplete: long (nullable = true)
 |    |    |-- incorrect: long (nullable = true)
 |    |    |-- submitted: long (nullable = true)
 |    |    |-- total: long (nullable = true)
 |    |    |-- unanswered: long (nullable = true)
 |    |-- id: string (nullable = true)
 |    |-- questions: array (nullable = true)
 |    |    |-- element: struct (containsNull = true)
 |    |    |    |-- id: string (nullable = true)
 |    |    |    |-- options: array (nullable = true)
 |    |    |    |    |-- element: struct (containsNull = true)
 |    |    |    |    |    |-- at: string (nullable = true)
 |    |    |    |    |    |-- checked: boolean (nullable = true)
 |    |    |    |    |    |-- correct: boolean (nullable = true)
 |    |    |    |    |    |-- id: 

#### We'll start by looking at 'sequences.attempt'

In [183]:
assessments_df.select('sequences.attempt').show(5)

+-------+
|attempt|
+-------+
|      1|
|      1|
|      1|
|      1|
|      1|
+-------+
only showing top 5 rows



In [96]:
assessments_df.select(countDistinct('sequences.attempt')).show()

+---------------------------------+
|count(DISTINCT sequences.attempt)|
+---------------------------------+
|                                1|
+---------------------------------+



In [94]:
assessments_df.select(F.sum('sequences.attempt')).collect()[0][0]

3280

#### All the values in 'sequences.attempt' are 1. I'm not sure this tells us much that is useful. 

#### We'll next look to 'sequences.counts'.

In [98]:
assessments_df.select("sequences.counts.correct", "sequences.counts.incorrect", \
                      "sequences.counts.incomplete","sequences.counts.unanswered",\
                      "sequences.counts.submitted", "sequences.counts.total" ).show()

+-------+---------+----------+----------+---------+-----+
|correct|incorrect|incomplete|unanswered|submitted|total|
+-------+---------+----------+----------+---------+-----+
|      2|        1|         1|         0|        4|    4|
|      1|        1|         2|         0|        4|    4|
|      3|        1|         0|         0|        4|    4|
|      2|        0|         2|         0|        4|    4|
|      3|        1|         0|         0|        4|    4|
|      5|        0|         0|         0|        5|    5|
|      1|        0|         0|         0|        1|    1|
|      5|        0|         0|         0|        5|    5|
|      4|        0|         0|         0|        4|    4|
|      0|        0|         1|         4|        1|    5|
|      3|        0|         1|         0|        4|    4|
|      1|        0|         0|         0|        1|    1|
|      4|        1|         1|         0|        6|    6|
|      4|        2|         0|         0|        6|    6|
|      4|     

In [195]:
#assessments_df.select("sequences.counts.*").show()

These are definitely interesting data and should be part of what's available to the analysts.
It would appear these are the test scores. 'correct' and 'total' in particular would seem to be what you'd need to calculate 
a percentage correct when evaluating an assessment result. 

#### We'll next look at 'sequences.id.*'

In [184]:
# are these the IDs of the questions taken?
assessments_df.select('sequences.questions.id').take(1)[0][0]


['7a2ed6d3-f492-49b3-b8aa-d080a8aad986',
 'bbed4358-999d-4462-9596-bad5173a6ecb',
 'e6ad8644-96b1-4617-b37b-a263dded202c',
 '95194331-ac43-454e-83de-ea8913067055']

In [186]:

# This seems to tie out the # of IDs to the total number of questions (below)
assessments_df.select(F.size('sequences.questions.id'), 'sequences.counts.total').show()

+----------------------------+-----+
|size(sequences.questions.id)|total|
+----------------------------+-----+
|                           4|    4|
|                           4|    4|
|                           4|    4|
|                           4|    4|
|                           4|    4|
|                           5|    5|
|                           1|    1|
|                           5|    5|
|                           4|    4|
|                           5|    5|
|                           4|    4|
|                           1|    1|
|                           6|    6|
|                           6|    6|
|                           5|    5|
|                           4|    4|
|                           4|    4|
|                           4|    4|
|                           4|    4|
|                           6|    6|
+----------------------------+-----+
only showing top 20 rows



#### 'sequences.questions.id' appears to be an array of question ID's. As a check, I compared the length of the array to the total question count. They appear to match up exactly. I would certainly keep this in the data passed to analysts as it could be used to join on unique question information. 

#### The last piece I'll tackle is 'sequences.questions', which is the most complex and most nested portion of the dataset. I'll only touch on it briefly. The key question I have w/r/t this data is what is in this portion of the dataset that is not reflected elsewhere? It appears to have per question, per assessment information. (i.i.e. Which particular questions on a particular assesment taken by a person were correct, incorrect, etc.?)  

In [111]:
assessments_df.select('sequences.questions').show(5)

+--------------------+
|           questions|
+--------------------+
|[[7a2ed6d3-f492-4...|
|[[95194331-ac43-4...|
|[[b9ff2e88-cf9d-4...|
|[[1f7c5def-904b-4...|
|[[620c924f-6bd8-1...|
+--------------------+
only showing top 5 rows



In [190]:
assessments_df.select('sequences.questions.options').take(1)[0][0][0][0]

Row(at='2018-01-23T14:23:24.670Z', checked=True, correct=True, id='49c574b4-5c82-4ffd-9bd1-c3358faf850d', submitted=1)

#### A fairly typical row appears above. It seems like this data might be redundant with some of the data captured in other fields. If I were to spend more time with this dataset, one thing I would want to make clearer is what do all the different id's mean. I'd relabel them to make them clearer. (i.e. assessment_id, question_id, etc.)

In [170]:
assessments_df.select('sequences').printSchema()

root
 |-- sequences: struct (nullable = true)
 |    |-- attempt: long (nullable = true)
 |    |-- counts: struct (nullable = true)
 |    |    |-- all_correct: boolean (nullable = true)
 |    |    |-- correct: long (nullable = true)
 |    |    |-- incomplete: long (nullable = true)
 |    |    |-- incorrect: long (nullable = true)
 |    |    |-- submitted: long (nullable = true)
 |    |    |-- total: long (nullable = true)
 |    |    |-- unanswered: long (nullable = true)
 |    |-- id: string (nullable = true)
 |    |-- questions: array (nullable = true)
 |    |    |-- element: struct (containsNull = true)
 |    |    |    |-- id: string (nullable = true)
 |    |    |    |-- options: array (nullable = true)
 |    |    |    |    |-- element: struct (containsNull = true)
 |    |    |    |    |    |-- at: string (nullable = true)
 |    |    |    |    |    |-- checked: boolean (nullable = true)
 |    |    |    |    |    |-- correct: boolean (nullable = true)
 |    |    |    |    |    |-- id: 


### QUERIES USING THE DATASET

### 1. How many assessments are in the dataset?

In [122]:
assessments_df.count()

3280

If each row is an assessment, then there were 3,280 assessments in the dataset. 

### 2. How many people took Learning Git?

In [148]:
assessments_df.filter(assessments_df["exam_name"] == "Learning Git").count()

394

394 people took 'Learning Git'

### 3. What is the least common course taken?

In [157]:
course_counts = assessments_df.select('exam_name').groupBy("exam_name").count()

In [158]:
min_course = course_counts.select('exam_name').groupBy("exam_name").count().agg({'count':'min'})

In [159]:
course_counts.filter(course_counts["count"] == min_course.collect()[0][0] ).show(course_counts.count(), False)

+-------------------------------------------------+-----+
|exam_name                                        |count|
+-------------------------------------------------+-----+
|Learning to Visualize Data with D3.js            |1    |
|Nulls, Three-valued Logic and Missing Information|1    |
|Native Web Apps for Android                      |1    |
|Operating Red Hat Enterprise Linux Servers       |1    |
+-------------------------------------------------+-----+



The 4 least commonly taken courses are above. 

### 4. What is the most commonly course taken?

In [163]:
course_counts.filter(course_counts["count"] == max_course.collect()[0][0] ).show(1, False)

+------------+-----+
|exam_name   |count|
+------------+-----+
|Learning Git|394  |
+------------+-----+



The most commonly taken course is above. 

### 5. How would I determine the score (= percent correct) for each assessment?

In [164]:
assessments_df.select( (  (col("sequences.counts.correct")/col("sequences.counts.total")).alias("pct_correct"))).show(10)

+-----------+
|pct_correct|
+-----------+
|        0.5|
|       0.25|
|       0.75|
|        0.5|
|       0.75|
|        1.0|
|        1.0|
|        1.0|
|        1.0|
|        0.0|
+-----------+
only showing top 10 rows



### 6. How many unique exams are offered?

In [165]:
assessments_df.select(countDistinct("exam_name")).show()



+-------------------------+
|count(DISTINCT exam_name)|
+-------------------------+
|                      103|
+-------------------------+



### 7. What percent of assessments are certified?

In [169]:

null_assess = assessments_df.filter(assessments_df["certification"].isNull()).select('certification').count()
false_assess = assessments_df.filter(assessments_df["certification"] == 'false').select('certification').count()
total_assess = assessments_df.count()

print("Number of assessments with certification == 'false' =", false_assess)
print("Number of assessments with certification == 'NULL' =", null_assess)
print("Total assessment =", total_assess)


Number of assessments with certification == 'false' = 3148
Number of assessments with certification == 'NULL' = 132
Total assessment = 3280


It turns out that the 'certification' field is either 'false' or NULL. So '0%' of assessments in the database are for certified assessments. In the ordinary course of work, I'd dig deeper to determine if this is a worthwhile field to keep. 


### WRITING TO HDFS

#### We'll redisplay the schema to discuss what we'll write to HDFS

In [197]:
assessments_df.printSchema()

root
 |-- base_exam_id: string (nullable = true)
 |-- certification: string (nullable = true)
 |-- exam_name: string (nullable = true)
 |-- keen_created_at: string (nullable = true)
 |-- keen_id: string (nullable = true)
 |-- keen_timestamp: string (nullable = true)
 |-- max_attempts: string (nullable = true)
 |-- sequences: struct (nullable = true)
 |    |-- attempt: long (nullable = true)
 |    |-- counts: struct (nullable = true)
 |    |    |-- all_correct: boolean (nullable = true)
 |    |    |-- correct: long (nullable = true)
 |    |    |-- incomplete: long (nullable = true)
 |    |    |-- incorrect: long (nullable = true)
 |    |    |-- submitted: long (nullable = true)
 |    |    |-- total: long (nullable = true)
 |    |    |-- unanswered: long (nullable = true)
 |    |-- id: string (nullable = true)
 |    |-- questions: array (nullable = true)
 |    |    |-- element: struct (containsNull = true)
 |    |    |    |-- id: string (nullable = true)
 |    |    |    |-- options: arra

#### We'll redisplay the schema to discuss what we'll write to HDFS

- I'll write all the top level nodes to one table and add a percent correct column
- I'll write all the members of sequences.counts to a separate table
- I'll write sequences.questions.id to a third table

For all the tables, I'll include user_exam_id as a key that they can be joined on. This is a little problematic because, as I mentioned earlier, this field isn't completely unique. In a real world environment I'd clean that column up or create a new unique id. 

For sequencies.questions.id, I'll write this as single column of arrays. With more time, I'd write this as three columns: user_exam_id, the sequence number of the item in the array and the id itself in a third column. 

In [216]:
top_level_df = assessments_df.select('user_exam_id', 'base_exam_id', 'certification', 'exam_name', 'keen_created_at', \
                                  'keen_id', 'keen_timestamp', 'max_attempts', 'started_at',
                                 (col("sequences.counts.correct")/col("sequences.counts.total")).alias("pct_correct")) 


In [218]:
count_df = assessments_df.select('user_exam_id', 'sequences.counts.*')

+--------------------+-----------+-------+----------+---------+---------+-----+----------+
|        user_exam_id|all_correct|correct|incomplete|incorrect|submitted|total|unanswered|
+--------------------+-----------+-------+----------+---------+---------+-----+----------+
|6d4089e4-bde5-4a2...|      false|      2|         1|        1|        4|    4|         0|
|2fec1534-b41f-441...|      false|      1|         2|        1|        4|    4|         0|
|8edbc8a8-4d26-429...|      false|      3|         0|        1|        4|    4|         0|
|c0ee680e-8892-4e6...|      false|      2|         2|        0|        4|    4|         0|
|e4525b79-7904-405...|      false|      3|         0|        1|        4|    4|         0|
|3186dafa-7acf-47e...|       true|      5|         0|        0|        5|    5|         0|
|48d88326-36a3-4cb...|       true|      1|         0|        0|        1|    1|         0|
|bb152d6b-cada-41e...|       true|      5|         0|        0|        5|    5|         0|

In [221]:
question_id_df = assessments_df.select('user_exam_id', 'sequences.questions.id')

In [222]:
question_id_df.show()

+--------------------+--------------------+
|        user_exam_id|                  id|
+--------------------+--------------------+
|6d4089e4-bde5-4a2...|[7a2ed6d3-f492-49...|
|2fec1534-b41f-441...|[95194331-ac43-45...|
|8edbc8a8-4d26-429...|[b9ff2e88-cf9d-4b...|
|c0ee680e-8892-4e6...|[1f7c5def-904b-48...|
|e4525b79-7904-405...|[620c924f-6bd8-11...|
|3186dafa-7acf-47e...|[fb07b16e-84a2-46...|
|48d88326-36a3-4cb...|[247b4589-7f8c-4a...|
|bb152d6b-cada-41e...|[fc3bdc54-04a8-4b...|
|70073d6f-ced5-4d0...|[803fc93f-7eb2-41...|
|9eb6d4d6-fd1f-4f3...|[fc3bdc54-04a8-4b...|
|093f1337-7090-457...|[a6effaf7-94ba-45...|
|0f576abb-958a-4c0...|[247b4589-7f8c-4a...|
|0c18f48c-0018-450...|[0d12c14d-1abe-4b...|
|b38ac9d8-eef9-495...|[26ddad33-aa1d-49...|
|bbc9865f-88ef-42e...|[7bdbbf4a-b5d8-4c...|
|8a0266df-02d7-44e...|[d2ac7f0d-82bd-41...|
|95d4edb1-533f-445...|[59d444b5-49fd-48...|
|f9bc1eff-7e54-42a...|[e272a3d1-bd67-4d...|
|dc4b35a7-399a-4bd...|[dee14932-a24e-4a...|
|d0f8249a-597e-4e1...|[861c3405-